rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java

```diff
@@ -611,44 +622,52 @@ private static <T> T last(List<T> l) {
         return l.get(l.size() - 1);
     }
 
-    private ILanguageContributions contributions(ISourceLocation doc) {
-        String language = registeredExtensions.get(extension(doc));
-        if (language != null) {
-            ILanguageContributions contrib = contributions.get(language);
+    private boolean isLanguageRegistered(ISourceLocation loc) {
+        return registeredExtensions.containsKey(extension(loc));
+    }
 
-            if (contrib != null) {
-                return contrib;
-            }
-        }
+    private Optional<String> safeLanguage(ISourceLocation loc) {
+        return Optional.ofNullable(registeredExtensions.get(extension(loc)));
+    }
 
-        throw new UnsupportedOperationException("Rascal Parametric LSP has no support for this file: " + doc);
+    private String language(ISourceLocation loc) {
+        return safeLanguage(loc).orElseThrow(() ->
+            new UnsupportedOperationException(String.format("Rascal Parametric LSP has no support for this file, since no language is registered for extension '%s': %s", extension(loc), loc))
+        );
+    }
+
+    private ILanguageContributions contributions(ISourceLocation doc) {
+        return safeLanguage(doc)
+            .map(contributions::get)
+            .map(ILanguageContributions.class::cast)
+            .flatMap(Optional::ofNullable)
+            .orElseGet(() -> new NoContributions(extension(doc)));
     }
 
     private static String extension(ISourceLocation doc) {
         return URIUtil.getExtension(doc);
     }
 
     private ParametricFileFacts facts(ISourceLocation doc) {
-        String language = registeredExtensions.get(extension(doc));
-        if (language != null) {
-            ParametricFileFacts fact = facts.get(language);
-            if (fact != null) {
-                return fact;
-            }
+        ParametricFileFacts fact = facts.get(language(doc));
+
+        if (fact == null) {
+            throw new ResponseErrorException(unknownFileError(doc, doc));
         }
-        throw new UnsupportedOperationException("Rascal Parametric LSP has no support for this file: " + doc);
+
+        return fact;
     }
 
     private TextDocumentState open(TextDocumentItem doc, long timestamp) {
-        return files.computeIfAbsent(Locations.toLoc(doc),
-            l -> new TextDocumentState(contributions(l)::parsing, l, doc.getVersion(), doc.getText(), timestamp)
-        );
+        var loc = Locations.toLoc(doc);
+        return files.computeIfAbsent(loc,
+            l -> new TextDocumentState(contributions(loc)::parsing, l, doc.getVersion(), doc.getText(), timestamp));
```

Comment:
`l` and `loc` are the same, no need to capture them from outside the closure.

(I don't see why we need this change?)